### PR TITLE
[fix](mtmv) Fix table id overturn and optimize get table qualifier method

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes

Table id may be the same but actually they are different tables. so we optimize the 
`org.apache.doris.nereids.rules.exploration.mv.mapping.RelationMapping#getTableQualifier` with following code:
`Objects.hash(table.getDatabase().getCatalog().getId(), table.getDatabase().getId(), table.getId())`

table id is long, we identify the table used in mv rewrite is bitSet. the bitSet can only use int, so we mapping the long id to init id in every query when mv rewrite

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/CascadesContext.java [hunk 1]
```diff
@@ -138,6 +142,10 @@ public class CascadesContext implements ScheduleContext {
     // into AggregateFunction with distinct, we can not fold constant in this case
     private int distinctAggLevel;
     private final boolean isEnableExprTrace;
+    // Record table id mapping, the key is the hash code of union catalogId, databaseId, tableId
+    // the value is the auto-increment id in the cascades context
+    private final Map<Long, Integer> tableIdMapping = new LinkedHashMap<>();
+    private final IdGenerator<TableId> talbeIdGenerator = TableId.createGenerator();
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/CascadesContext.java:148]: Use class like `TableIdentifier`

Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/CascadesContext.java:148]: put into StatementContext

File: fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java [hunk 2]
```diff
@@ -500,7 +500,7 @@ public String getExplainString(ExplainOptions explainOptions) {
                         this.getPhysicalPlan());
                 if (statementContext != null) {
                     if (statementContext.isHasUnknownColStats()) {
-                        plan += "planed with unknown column statistics\n";
+                        plan += "\n Statistics \n planed with unknown column statistics\n";
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java:503]:
```suggestion
                        plan += "\n\nStatistics\n  planned with unknown column statistics\n";
```

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Review by keanji-x — APPROVED** [review]

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/StatementContext.java [hunk 3]
```diff
@@ -141,6 +146,9 @@ public class StatementContext implements Closeable {
     // and value is the new string used for replacement.
     private final TreeMap<Pair<Integer, Integer>, String> indexInSqlToString
             = new TreeMap<>(new Pair.PairComparator<>());
+    // Record table id mapping, the key is the hash code of union catalogId, databaseId, tableId
+    // the value is the auto-increment id in the cascades context
+    private final Map<TableIdentifier, Integer> tableIdMapping = new LinkedHashMap<>();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/StatementContext.java:151]:
```suggestion
    private final Map<TableIdentifier, TableId> tableIdMapping = new LinkedHashMap<>();
```

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/MaterializationContext.java [hunk 4]
```diff
@@ -78,7 +78,7 @@ public abstract class MaterializationContext {
     // Mark enable record failure detail info or not, because record failure detail info is performance-depleting
     protected final boolean enableRecordFailureDetail;
     // The mv plan struct info
-    protected final StructInfo structInfo;
+    protected StructInfo structInfo;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/MaterializationContext.java:81]: why remove final?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/StructInfo.java [hunk 5]
```diff
@@ -395,6 +402,7 @@ public Map<ExprId, Expression> getNamedExprIdAndExprMapping() {
     }
 
     public BitSet getTableBitSet() {
+
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/StructInfo.java:405]: useless blank line

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/mapping/RelationMapping.java [hunk 6]
```diff
@@ -142,7 +143,7 @@ public static RelationMapping merge(List<BiMap<MappedRelation, MappedRelation>>
     }
 
     private static Long getTableQualifier(TableIf tableIf) {
-        return tableIf.getId();
+        return (long) new TableIdentifier(tableIf).hashCode();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/mapping/RelationMapping.java:146]: use TableIdentifier directly

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/MaterializationContext.java (hunk 4 above)
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/MaterializationContext.java:81]: construct sturctInfo may be failed when materialization init as following:
field `StructInfo` maybe not init, so remove the final
```
 List<StructInfo> viewStructInfos;
        try {
            viewStructInfos = MaterializedViewUtils.extractStructInfo(mvPlan, cascadesContext, new BitSet());
            if (viewStructInfos.size() > 1) {
                // view struct info should only have one, log error and use the first struct info
                LOG.warn(String.format("view strut info is more than one, materialization name is %s, mv plan is %s",
                        getMaterializationQualifier(), getMvPlan().treeString()));
            }
        } catch (Exception exception) {
            LOG.warn(String.format("construct mv struct info fail, materialization name is %s, mv plan is %s",
                    getMaterializationQualifier(), getMvPlan().treeString()), exception);
            this.available = false;
            return;
        }
        this.structInfo = viewStructInfos.get(0);
```

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/MaterializationContext.java (hunk 4 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/MaterializationContext.java:81]: set structInfo to null before return in catch?

**Comment by seawinde** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Pull request merged**